mixer/adapter/stackdriver/contextgraph/contextgraph.go

```diff
@@ -0,0 +1,189 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package contextgraph adapter for Stackdriver Context API.
+package contextgraph
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	contextgraph "cloud.google.com/go/contextgraph/apiv1alpha1"
+	gax "github.com/googleapis/gax-go"
+
+	"google.golang.org/api/option"
+	contextgraphpb "google.golang.org/genproto/googleapis/cloud/contextgraph/v1alpha1"
+	"istio.io/istio/mixer/adapter/stackdriver/config"
+	"istio.io/istio/mixer/adapter/stackdriver/helper"
+	"istio.io/istio/mixer/pkg/adapter"
+	edgepb "istio.io/istio/mixer/template/edge"
+)
+
+type newClientFn func(context.Context, ...option.ClientOption) (*contextgraph.Client, error)
+type assertBatchFn func(context.Context, *contextgraphpb.AssertBatchRequest, ...gax.CallOption) (*contextgraphpb.AssertBatchResponse, error)
+
+type (
+	builder struct {
+		projectID string
+		zone      string
+		cluster   string
+		mg        helper.MetadataGenerator
+		opts      []option.ClientOption
+		newClient newClientFn
+	}
+	handler struct {
+		client         *contextgraph.Client
+		env            adapter.Env
+		projectID      string
+		meshUID        string
+		zone, cluster  string
+		entityCache    *entityCache
+		edgeCache      *edgeCache
+		traffics       chan trafficAssertion
+		entitiesToSend []entity
+		edgesToSend    []edge
+		sendTick       *time.Ticker
+		quit           chan int
+		assertBatch    assertBatchFn
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ edgepb.HandlerBuilder = &builder{}
+var _ edgepb.Handler = &handler{}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+
+	env.Logger().Debugf("Proj, zone, cluster, opts: %s,%s,%s,%s",
+		b.projectID, b.zone, b.cluster, b.opts)
+
+	// TODO: meshUID should come from an attribute when
+	// multi-cluster Istio is supported. Currently we assume each
+	// cluster is its own mesh.
+	h := &handler{
+		env:            env,
+		projectID:      b.projectID,
+		meshUID:        fmt.Sprintf("%s/%s/%s", b.projectID, b.zone, b.cluster),
+		zone:           b.zone,
+		cluster:        b.cluster,
+		entityCache:    newEntityCache(env.Logger()),
+		edgeCache:      newEdgeCache(env.Logger()),
+		traffics:       make(chan trafficAssertion),
+		entitiesToSend: make([]entity, 0),
+		edgesToSend:    make([]edge, 0),
+		sendTick:       time.NewTicker(30 * time.Second),
```

Comment:
should `30s` be a const, and/or configurable?

Reply:
I don't think it should be user configurable, but maybe. The ContextGraph API has a very low quota, and expects us to batch up asserts.